# CSC 593

## Week 7

### Merge Errors

#### Not resolved automatically
`git checkout --ours PATH/FILE`

#### Resolved automatically (vim)
`:wq`

### Pandas 2

In [2]:
import numpy as np
import pandas as pd

The BRFSS (Behavioral Risk Factor Surveillance System) data is too big to put into Github. This cell downloads it from the CDC's website and unzips it into your `data` folder.

In [3]:
#Setup for examples.
from urllib.request import urlretrieve
import zipfile
from pathlib import Path

zf = '../data/brfss/LLCP2018ASC.zip'
if not Path(zf).exists():
    Path('../data/brfss').mkdir(exist_ok=True)

    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/pdf/overview-2018-508.pdf', '../data/brfss/overview-2018-508.pdf')
    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/pdf/codebook18_llcp-v2-508.pdf', '../data/brfss/codebook18_llcp-v2-508.pdf')
    
    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/files/LLCP2018ASC.zip', zf)

fwff = '../data/brfss/LLCP2018.ASC'
if not Path(fwff).exists():
    with zipfile.ZipFile(zf) as z:
        z.extractall('../data/brfss')

Load the BRFSS data and set a couple of data types explicitly. (More supported data types are listed at https://docs.scipy.org/doc/numpy/user/basics.types.html)

In [4]:
names= ['state', 'imonth', 'iday', 
        'iyear', 'dispcode','genhlth', 
        'physhlth',
        'menthlth', 'poorhlth', 'hlthpln1',
        'persdoc2', 'medcost', 'checkup1',
        'WEIGHT2', 'HEIGHT3']
cols = [
    (1, 3),
    (18, 20),
    (20, 22),
    (22, 27),
    (31, 35),
    (89, 90),
    (90, 92),
    (92, 94),
    (94, 96),
    (96, 97),
    (97, 98),
    (98, 99),
    (99, 100),
    (176, 180),
    (180, 184)
]
types= {
    'WEIGHT2': str, 
    'HEIGHT3': str,
}
brfss = pd.read_fwf(fwff + ' ',
                    names=names,
                    colspecs=cols,
                    dtype=types)

#### Searching

In [5]:
#Get an individual column.
brfss['WEIGHT2']

0         0130
1         0200
2         0142
3         0190
4         7777
5         0172
6         0150
7         0130
8         0205
9         0151
10        0199
11        0170
12        0160
13        0200
14        0285
15        0217
16        0230
17        0250
18        0155
19        0138
20        0240
21        0240
22        0185
23        0165
24        0131
25        0119
26        0180
27        0169
28        0128
29        0218
          ... 
428016    0195
428017    0140
428018    0150
428019    0270
428020    9999
428021    0125
428022    0206
428023    0150
428024    0200
428025    0210
428026    0175
428027    0162
428028    0230
428029    0250
428030    0220
428031     NaN
428032    0160
428033    0153
428034    0160
428035    0130
428036    0148
428037    0152
428038    0135
428039    0123
428040    0165
428041    0155
428042    0130
428043    0230
428044    0141
428045     NaN
Name: WEIGHT2, Length: 428046, dtype: object

In [6]:
#For multiple columns, use a list as a subscript.
brfss[['WEIGHT2', 'HEIGHT3']]

WEIGHT2 HEIGHT3
0         0130    0504
1         0200    0505
2         0142    0410
3         0190    0510
4         7777    0503
5         0172    0502
6         0150    0505
7         0130    0505
8         0205    0601
9         0151    0503
10        0199    0600
11        0170    0503
12        0160    0503
13        0200    0504
14        0285    0511
15        0217    0506
16        0230    0509
17        0250    0601
18        0155    0504
19        0138    0504
20        0240    0507
21        0240    0504
22        0185    0505
23        0165    0509
24        0131    0504
25        0119    0502
26        0180    0505
27        0169    0506
28        0128    0506
29        0218    0603
...        ...     ...
428016    0195    0602
428017    0140    0505
428018    0150    0504
428019    0270    0509
428020    9999    0508
428021    0125    0506
428022    0206    0600
428023    0150    0505
428024    0200    0510
428025    0210    0600
428026    0175    0601
428027    0162    0505
428028    0230    0506
428029    0250    0510
428030    0220    0510
428031     NaN     NaN
428032    0160    0502
428033    0153    0504
428034    0160    0507
428035    0130    0504
428036    0148    0506
428037    0152    0509
428038    0135    0500
428039    0123    0502
428040    0165    0502
428041    0155    0508
428042    0130    0506
428043    0230    0506
428044    0141    0506
428045     NaN     NaN

[428046 rows x 2 columns]

The `loc()` and `iloc()` methods (see the table on p. 144-5 of *Python for Data Analysis*)

In [7]:
#Get the first row.
brfss.loc[0]

state          1
imonth         1
iday           5
iyear       2018
dispcode    1100
genhlth        2
physhlth      30
menthlth      88
poorhlth      30
hlthpln1       1
persdoc2       1
medcost        2
checkup1       1
WEIGHT2     0130
HEIGHT3     0504
Name: 0, dtype: object

In [8]:
#Get WEIGHT2 from the third row.
brfss.loc[2, 'WEIGHT2']

'0142'

In [9]:
#Same thing, but using the integer index instead of the column name.
brfss.iloc[2,-2]

'0142'

In [10]:
brfss.at[2, 'WEIGHT2']

'0142'

In [11]:
brfss.iat[2, -2]

'0142'

In [12]:
brfss.loc[:100, ['HEIGHT3', 'WEIGHT2']]

HEIGHT3 WEIGHT2
0      0504    0130
1      0505    0200
2      0410    0142
3      0510    0190
4      0503    7777
5      0502    0172
6      0505    0150
7      0505    0130
8      0601    0205
9      0503    0151
10     0600    0199
11     0503    0170
12     0503    0160
13     0504    0200
14     0511    0285
15     0506    0217
16     0509    0230
17     0601    0250
18     0504    0155
19     0504    0138
20     0507    0240
21     0504    0240
22     0505    0185
23     0509    0165
24     0504    0131
25     0502    0119
26     0505    0180
27     0506    0169
28     0506    0128
29     0603    0218
..      ...     ...
71     0503    7777
72     0506    0150
73     0506    0160
74     0601    0232
75     0508    0222
76     0510    0210
77     0500    0170
78     0500    0237
79     0505    0188
80     0505    0125
81     0506    0180
82     0509    0180
83     0500    0160
84     0507    0225
85     0506    0160
86     0506    0160
87     0500    0120
88     0510    0228
89     0508    0205
90     0502    9999
91     0600    0235
92     0502    0142
93     0511    0210
94     0509    0182
95     0508    0220
96     0410    0125
97     0600    0190
98     0600    0200
99     0507    0170
100    0510    0200

[101 rows x 2 columns]

In [13]:
#Find rows based on a value
brfss[brfss['WEIGHT2']=='9999']

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
61          1       2     1   2018      1100      1.0      88.0      88.0   
69          1       2     1   2018      1100      4.0      30.0      88.0   
90          1       2     3   2018      1100      1.0      88.0      14.0   
118         1       3     5   2018      1100      2.0      88.0      88.0   
129         1       3    15   2018      1100      1.0      88.0      88.0   
137         1       1    13   2018      1100      2.0      88.0      88.0   
159         1       1    11   2018      1100      2.0      88.0      88.0   
162         1       1     8   2018      1100      3.0      88.0      14.0   
172         1       1    25   2018      1100      2.0      88.0      88.0   
191         1       2     3   2018      1100      4.0      30.0      88.0   
231         1       3     1   2018      1100      5.0      30.0      15.0   
234         1       3    12   2018      1100      4.0      20.0      10.0   
271         1       1    11   2018      1100      3.0      88.0      88.0   
296         1       1     8   2018      1100      3.0      88.0      88.0   
320         1       3     1   2018      1100      2.0      10.0      88.0   
324         1       2     2   2018      1200      2.0      88.0      88.0   
347         1       3    24   2018      1100      2.0      88.0      88.0   
465         1       3     2   2018      1100      4.0      15.0      88.0   
475         1       3    22   2018      1100      3.0      88.0      88.0   
529         1       3     6   2018      1200      3.0      99.0      88.0   
534         1       2    14   2018      1100      5.0      30.0      88.0   
573         1       3    28   2018      1100      2.0      88.0      88.0   
590         1       1    11   2018      1100      3.0       3.0       7.0   
678         1       3     2   2018      1100      3.0      88.0      88.0   
694         1       1    23   2018      1200      2.0       7.0       5.0   
719         1       2    27   2018      1100      5.0      20.0      15.0   
728         1       2    15   2018      1100      1.0      88.0      88.0   
744         1       3     1   2018      1100      3.0      15.0      88.0   
773         1       3    22   2018      1100      2.0      88.0      88.0   
777         1       1    22   2018      1100      3.0      88.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
427268      6       5    10   2018      1100      1.0      88.0      88.0   
427284      6       5    10   2018      1100      4.0      88.0      88.0   
427339      6       5     9   2018      1100      3.0      30.0      88.0   
427366      6       6    17   2018      1100      2.0      88.0      88.0   
427373      6       6    11   2018      1100      3.0      88.0       4.0   
427383      6       7     2   2018      1100      3.0       7.0       7.0   
427405      6       6    13   2018      1100      3.0      88.0      88.0   
427419      6       6    13   2018      1100      1.0      88.0      88.0   
427460      6       6    28   2018      1100      2.0      88.0      88.0   
427470      6       6    14   2018      1100      2.0      88.0      88.0   
427483      6       6    14   2018      1100      4.0       7.0       4.0   
427523      6       6    15   2018      1100      3.0      88.0      99.0   
427525      6       6    21   2018      1100      2.0      88.0       4.0   
427564      6       7     8   2018      1200      2.0      88.0      88.0   
427568      6       7    10   2018      1100      4.0      30.0      88.0   
427646      6       7    15   2018      1200      2.0      88.0      88.0   
427713      6       7    15   2018      1100      1.0      88.0      88.0   
427717      6       7    28   2018      1100      3.0       1.0       1.0   
427755      6       8     1   2018      1100      3.0       4.0       4.0   
427776      6       8     1   2018      1100      2.0      88.0      88.0   
4277

The [`shape()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html) method gives you the height and width of your DataFrame.

In [14]:
print(brfss.shape)
#Drop any rows without weight
brfss.dropna(subset=['WEIGHT2'], inplace=True)
print(brfss.shape)

(428046, 15)
(421697, 15)


https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.isin.html

In [15]:
brfss[np.isin(brfss['WEIGHT2'], ['7777', '9999'], invert=True)].shape

(399857, 15)

#### Derived Fields

In [16]:
brfss = brfss[np.isin(brfss['WEIGHT2'], ['7777', '9999'], invert=True)]
brfss = brfss[~brfss.WEIGHT2.str.startswith('1')]

In [17]:
brfss['wtunit'] = brfss.WEIGHT2.str[0].astype(np.uint8)
brfss['wt'] = brfss.WEIGHT2.str[1:].astype(np.uint16)
brfss

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
428013      6       9     5   2018      1100      5.0      30.0      88.0   
428014      6       9     8   2018      1100      2.0      88.0      88.0   
428015      6       9    30   2018      1100      1.0      88.0      88.0   
428016      6       9    21   2018      1100      1.0      88.0      88.0   
428017      6       9     5   2018      1100      1.0      88.0      88.0   
428018      6       9    28   2018      1100      2.0      88.0      88.0   
428019      6      10     2   2018      1100      4.0      88.0      88.0   
428021      6       9     2   2018      1100      3.0      88.0      88.0   
428022      6       9     2   2018      1100      3.0      30.0      88.0   
428023      6       9     5   2018      1100      2.0      88.0      88.0   
428024      6       9     4   2018      1100      4.0      10.0       2.0   
428025      6       9     2   2018      1100      4.0       5.0      88.0   
428026      6       9     7   2018      1100      3.0      88.0      88.0   
428027      6       9     2   2018      1100      3.0      88.0      88.0   
428028      6       9    19   2018      1100      4.0      30.0      88.0   
428029      6       9    16   2018      1100      5.0      30.0       2.0   
428030      6       9     6   2018      1100      4.0      88.0      88.0   
428032      6       9     6   2018      1100      3.0      88.0      88.0   
428033      6       9     7   2018      1100      2.0      88.0      88.0   
428034      6       9     7   2018      1100      2.0      88.0      88.0   
4280

[`np.where()`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.where.html) provides the equivalent of an *if-then-else* statement on each observation in a DataFrame:

In [18]:
lbsperkg = 2.205
brfss['wtlbs'] = np.where(brfss.wtunit==9, brfss.wt*lbsperkg, brfss.wt).astype(np.int16)

In [19]:
brfss[brfss.wtunit==9]

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
2732        1       3    27   2018      1100      3.0      88.0      88.0   
5182        1       9    26   2018      1100      2.0      88.0      88.0   
8442        2       6     4   2018      1100      4.0      30.0      30.0   
9092        2      10    30   2018      1100      2.0       1.0      88.0   
10312       4       8     7   2018      1100      2.0      14.0      88.0   
10404       4      10     4   2018      1100      4.0      88.0      88.0   
10780       4       1    24   2018      1100      3.0      88.0      88.0   
11404       4       7     9   2018      1100      2.0      88.0      88.0   
11773       4      12     2   2018      1100      2.0      88.0      88.0   
13098       4       2    15   2018      1100      1.0      88.0      88.0   
13124       4       2     7   2018      1100      1.0      88.0      88.0   
13176       4       2    11   2018      1100      1.0      88.0      88.0   
13199       4       2    11   2018      1100      4.0      88.0      88.0   
13357       4       3    30   2018      1100      1.0      88.0       3.0   
13380       4       3    18   2018      1100      4.0       3.0      77.0   
13759       4       4    13   2018      1100      1.0      88.0      88.0   
13960       4       7     3   2018      1200      4.0       2.0      88.0   
14365       4       8     8   2018      1100      3.0      88.0      88.0   
14366       4       8     9   2018      1100      2.0      88.0      88.0   
14415       4       8     7   2018      1100      1.0      88.0      88.0   
14599       4      10     1   2018      1100      4.0      15.0      88.0   
14640       4      10    23   2018      1100      4.0       2.0       1.0   
14844       4      10    18   2018      1100      3.0       2.0       1.0   
14896       4      11    19   2018      1100      1.0      88.0      88.0   
15001       4      11     7   2018      1100      5.0      30.0      30.0   
15073       4      12    27   2018      1100      5.0       7.0      88.0   
15672       4       3    18   2018      1100      4.0       6.0       6.0   
16451       4       6    15   2018      1100      3.0      15.0      88.0   
16481       4       7    18   2018      1100      4.0      88.0      88.0   
16905       4       9     8   2018      1100      2.0      88.0      15.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
415855      3      11     6   2018      1100      1.0      88.0      88.0   
415868      3      11    12   2018      1100      3.0       2.0      88.0   
415930      3      11    16   2018      1100      3.0      88.0      88.0   
415952      3      12    21   2018      1200      2.0       4.0      88.0   
415981      3      12    21   2018      1100      1.0      88.0       7.0   
416113      3      12     7   2018      1100      4.0      77.0      88.0   
416302      3      11    14   2018      1100      4.0      77.0      77.0   
416339      3      11    11   2018      1100      2.0      88.0      88.0   
416475      3      12    13   2018      1100      1.0      88.0      88.0   
416488      3      12    22   2018      1100      1.0      88.0      88.0   
419155      4       1    18   2018      1100      1.0      88.0      88.0   
419961      4       5    22   2018      1100      5.0      30.0      88.0   
421203      4      11    18   2018      1100      1.0      30.0       2.0   
421788      5       2    18   2018      1200      3.0      77.0      15.0   
422061      5       4    10   2018      1100      2.0      88.0      88.0   
422222      5       5    11   2018      1100      3.0      88.0      88.0   
423401      5       8    13   2018      1200      2.0      88.0      88.0   
423440      5       1    11   2019      1200      3.0      88.0      88.0   
424002      5       3     1   2018      1100      1.0      88.0      88.0   
424296      5       3    21   2018      1100      3.0      88.0      88.0   
4243

##### Practice

Create a `htinches` column from the `brfss.HEIGHT3` column. 

1. Remove rows where `HEIGHT3` is 7777 ("Don't know/Not sure"),  9999 ("Refused"), or NaN.
2. If the first character of `HEIGHT3` is '9', multiply the remaining three digits by `cmtoin` (defined below) to get height in inches.
3. If the first character of `HEIGHT3` is '0', the second character is feet, and the third and fourth are inches ('0601' means six feet, one inch). Convert this to inches.

See page 36 of the codebook for details on the `HEIGHT3` field.

In [ ]:
cmtoin = 0.3937

In [23]:
brfss.dropna(subset=['HEIGHT3'], inplace=True)

brfss['htunit']=brfss.HEIGHT3.str[0].astype(np.uint16)
brfss['ht'] = brfss.HEIGHT3.str[1:]
brfss.ht
feet = brfss.ht.str[0].astype(np.uint8)
inches = brfss.ht.str[1:].astype(np.uint8)
inches = inches + (feet *12)
brfss['htinches'] = np.where(brfss.htunit==0, inches, brfss.ht)


#### Summary statistics and aggregation

In [24]:
brfss.groupby(['persdoc2', 'poorhlth']).size() #or .mean()

persdoc2  poorhlth
1.0       1.0          7710
          2.0          8777
          3.0          5599
          4.0          3157
          5.0          6133
          6.0          1003
          7.0          2982
          8.0           746
          9.0           174
          10.0         5170
          11.0           59
          12.0          490
          13.0           61
          14.0         1598
          15.0         5395
          16.0          131
          17.0           84
          18.0          159
          19.0           17
          20.0         3511
          21.0          452
          22.0           66
          23.0           45
          24.0           71
          25.0         1349
          26.0           67
          27.0           87
          28.0          272
          29.0          101
          30.0        13844
                      ...  
7.0       14.0            6
          15.0           22
          18.0            2
          20.0           16
 

In [25]:
# | means 'or'
# ph=1 if you were sick more than 5 days, 0 otherwise:
brfss['ph'] = np.where((brfss['poorhlth'] > 30) | (brfss['poorhlth'] <= 5) | (brfss['poorhlth'].isnull()), 0, 1)
brfss

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
428013      6       9     5   2018      1100      5.0      30.0      88.0   
428014      6       9     8   2018      1100      2.0      88.0      88.0   
428015      6       9    30   2018      1100      1.0      88.0      88.0   
428016      6       9    21   2018      1100      1.0      88.0      88.0   
428017      6       9     5   2018      1100      1.0      88.0      88.0   
428018      6       9    28   2018      1100      2.0      88.0      88.0   
428019      6      10     2   2018      1100      4.0      88.0      88.0   
428021      6       9     2   2018      1100      3.0      88.0      88.0   
428022      6       9     2   2018      1100      3.0      30.0      88.0   
428023      6       9     5   2018      1100      2.0      88.0      88.0   
428024      6       9     4   2018      1100      4.0      10.0       2.0   
428025      6       9     2   2018      1100      4.0       5.0      88.0   
428026      6       9     7   2018      1100      3.0      88.0      88.0   
428027      6       9     2   2018      1100      3.0      88.0      88.0   
428028      6       9    19   2018      1100      4.0      30.0      88.0   
428029      6       9    16   2018      1100      5.0      30.0       2.0   
428030      6       9     6   2018      1100      4.0      88.0      88.0   
428032      6       9     6   2018      1100      3.0      88.0      88.0   
428033      6       9     7   2018      1100      2.0      88.0      88.0   
428034      6       9     7   2018      1100      2.0      88.0      88.0   
4280

In [26]:
#brfss.groupby(['persdoc2', 'ph']).describe()
brfss.ph.groupby(brfss.persdoc2).size()

persdoc2
1.0    300187
2.0     30776
3.0     66090
7.0      1258
9.0       297
Name: ph, dtype: int64

In [27]:
#Calculate percentages instead of raw numbers.
docph = brfss.groupby(['persdoc2', 'ph']).size()
docph.groupby(level=0).apply(lambda x: 100 * x / float(x.sum()))

persdoc2  ph
1.0       0     87.363210
          1     12.636790
2.0       0     80.423057
          1     19.576943
3.0       0     89.617189
          1     10.382811
7.0       0     87.758347
          1     12.241653
9.0       0     86.531987
          1     13.468013
dtype: float64

We can bin or categorize numeric variables with [`pd.cut()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html).

In [28]:
## 88 means 'none'; want to bin it separately from "Don't know" and "refused"
brfss.loc[brfss.poorhlth==88, 'poorhlth']=51
bins = [1, 5, 10, 20, 30, 51, 70]
brfss['phcats'] = pd.cut(brfss.poorhlth, bins, 
                         labels=['less than 5', 'less than 10', 
                                 'less than 20', 'more than 20', 'none', "don't know/refused"])
brfss

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
428013      6       9     5   2018      1100      5.0      30.0      88.0   
428014      6       9     8   2018      1100      2.0      88.0      88.0   
428015      6       9    30   2018      1100      1.0      88.0      88.0   
428016      6       9    21   2018      1100      1.0      88.0      88.0   
428017      6       9     5   2018      1100      1.0      88.0      88.0   
428018      6       9    28   2018      1100      2.0      88.0      88.0   
428019      6      10     2   2018      1100      4.0      88.0      88.0   
428021      6       9     2   2018      1100      3.0      88.0      88.0   
428022      6       9     2   2018      1100      3.0      30.0      88.0   
428023      6       9     5   2018      1100      2.0      88.0      88.0   
428024      6       9     4   2018      1100      4.0      10.0       2.0   
428025      6       9     2   2018      1100      4.0       5.0      88.0   
428026      6       9     7   2018      1100      3.0      88.0      88.0   
428027      6       9     2   2018      1100      3.0      88.0      88.0   
428028      6       9    19   2018      1100      4.0      30.0      88.0   
428029      6       9    16   2018      1100      5.0      30.0       2.0   
428030      6       9     6   2018      1100      4.0      88.0      88.0   
428032      6       9     6   2018      1100      3.0      88.0      88.0   
428033      6       9     7   2018      1100      2.0      88.0      88.0   
428034      6       9     7   2018      1100      2.0      88.0      88.0   
4280

##### Practice

Create a new column that divides `iday` into 3 bins (1-10, 11-20, 21+).

In [ ]:
brfss.loc[brfss.poorhlth==88, 'poorhlth']=51
bins = [1, 5, 10, 20, 30, 51, 70]
brfss['phcats'] = pd.cut(brfss.poorhlth, bins, 
                         labels=['less than 5', 'less than 10', 
                                 'less than 20', 'more than 20', 'none', "don't know/refused"])

2) Group `brfss` by `hlthpln1` and `medcost` and create a table like the one above (for `persdoc2` and `ph`) with percentages for each subgroup.